tesseract-server/src/logic_layer/cache.rs

```diff
@@ -0,0 +1,194 @@
+            if cube_cache.name == *cube {
+                return Some(cube_cache.clone());
+            }
+        }
+        None
+    }
+}
+
+/// Holds cache information for a given cube.
+#[derive(Debug, Clone)]
+pub struct CubeCache {
+    pub name: String,
+    pub time_dim: Dimension,
+    pub years: Vec<u32>,
+}
+
+impl CubeCache {
+    /// Returns dimension name in the format: `Dimension.Hierarchy.Level`.
+    pub fn get_time_level_name(&self) -> LevelName {
+        LevelName {
+            dimension: self.time_dim.name.clone(),
+            hierarchy: self.time_dim.hierarchies[0].name.clone(),
+            level: self.time_dim.hierarchies[0].levels[0].name.clone(),
+        }
+    }
+
+    pub fn get_time_cut(&self, t: Time) -> Result<String, Error> {
+        let year_opt;
+
+        // TODO: Add check for precision type
+        match t.value {
+            TimeValue::Last => year_opt = self.max_year(),
+            TimeValue::First => year_opt = self.min_year(),
+            TimeValue::Value(time) => year_opt = Some(time),
+        }
+
+        let year = match year_opt {
+            None => { return Err(format_err!("Unable to get requested year.")); }
+            Some(year) => year
+        };
+
+        let ln = self.get_time_level_name();
+
+        Ok(format!("[{}].[{}].[{}].[{}]", ln.dimension(), ln.hierarchy(), ln.level(), year).to_string())
+    }
+
+    pub fn min_year(&self) -> Option<u32> {
+        if self.years.len() >= 1 {
+            return Some(self.years[0]);
+        }
+        None
+    }
+
+    pub fn max_year(&self) -> Option<u32> {
+        if self.years.len() >= 1 {
+            return Some(*self.years.last().unwrap());
+        }
+        None
+    }
+}
+
+
+/// Populates a `Cache` object that will be shared through `AppState`.
+pub fn populate_cache(schema: Schema, backend: Box<dyn Backend + Sync + Send>) -> Result<Cache, Error> {
+    info!("Populating cache...");
+
+    let mut sys = actix::System::new("cache");
+    let mut cubes: Vec<CubeCache> = vec![];
+
+    for cube in schema.cubes {
+        let preferred_time_dim = match find_years(cube.clone()) {
+            Ok(r) => match r {
+                Some(dim) => dim,
+                None => continue
+            },
+            Err(_) => continue
+        };
+        let year_column = get_year_column(&preferred_time_dim);
+
+        let future = backend
+            .exec_sql(
+                format!("select distinct {} from {}", year_column, cube.table.name)
+                        .to_string()
+            );
+        let df = match sys.block_on(future) {
+            Ok(df) => df,
+            Err(err) => {
+                return Err(format_err!("Error populating cache with backend data: {}", err));
+            }
+        };
+
+        // TODO: Do we want to return an error if no columns are returned?
+        if df.columns.len() >= 1 {
+            let mut original_years = match &df.columns[0].column_data {
```

Comment:
Just a heads up that I'll be adding null support in the near future, which will add some more `ColumnData` types like `NullableInt8`, etc.

Reply:
I need to refactor this file anyway. I'll wait until then.